src/promptflow/promptflow/_sdk/_submitter/experiment_orchestrator.py

```diff
@@ -127,91 +158,542 @@ def _test_flow_node(self, node, test_context):
-            command=node.command,
-            environment_variables={**node.environment_variables, **node_context},
-            config=Configuration(overrides={Configuration.RUN_OUTPUT_PATH: run_output_path}),
+    def _resolve_column_mapping(self):
+        """Resolve column mapping with experiment inputs to constant values."""
+        logger.info(f"Start resolve node {self.node.name!r} column mapping.")
+        resolved_mapping = {}
+        for name, value in self.column_mapping.items():
+            if not isinstance(value, str) or not value.startswith("${inputs."):
+                resolved_mapping[name] = value
+                continue
+            input_name = value.split(".")[1].replace("}", "")
+            if input_name not in self.experiment_inputs:
+                raise ExperimentValueError(
+                    f"Node {self.node_name!r} inputs {value!r} related experiment input {input_name!r} not found."
+                )
+            resolved_mapping[name] = self.experiment_inputs[input_name].default
+        logger.debug(f"Resolved node {self.node.name!r} column mapping {resolved_mapping}.")
+        self.column_mapping = resolved_mapping
+
+    def _resolve_input_dirs(self):
+        logger.info("Start resolve node %s input dirs.", self.node.name)
+        # Get the node referenced data and run
+        referenced_data, referenced_run = ExperimentHelper.get_referenced_data_and_run(
+            node_name=self.node.name,
+            column_mapping=self.column_mapping,
+            experiment_data=self.experiment_data,
+            experiment_runs=self.node_runs,
+        )
+        if len(referenced_data) > 1:
+            raise ExperimentValueError(
+                f"Experiment flow node {self.node.name!r} has multiple data inputs {referenced_data}, "
+                "only 1 is expected."
+            )
+        if len(referenced_run) > 1:
+            raise ExperimentValueError(
+                f"Experiment flow node {self.node.name!r} has multiple run inputs {referenced_run}, "
+                "only 1 is expected."
+            )
+        (data_name, data_obj) = next(iter(referenced_data.items())) if referenced_data else (None, None)
+        (run_name, run_obj) = next(iter(referenced_run.items())) if referenced_run else (None, None)
+        logger.debug(f"Resolve node {self.node.name} referenced data {data_name!r}, run {run_name!r}.")
+        # Build inputs from experiment data and run
+        result = {}
+        if data_obj:
+            result.update({f"data.{data_name}": data_obj.path})
+        if run_obj:
+            result.update(ExperimentHelper.resolve_binding_from_run(run_name, run_obj, self.run_operations))
+        result = {k: str(Path(v).resolve()) for k, v in result.items() if v is not None}
+        logger.debug(f"Resolved node {self.node.name} input dirs {result}.")
+        return result
+
+    def _calculate_snapshot(self):
+        def calculate_files_content_hash(file_path):
```

Comment:
may extract snapshot related logic in another file

Reply:
extract snapshot logic  to utils.py